contracts/exchange/test/internal.ts

```diff
@@ -48,16 +50,358 @@ blockchainTests('Exchange core internal functions', env => {
+            feeRecipientAddress: randomAddress(),
+            domain,
+        };
+        const privateKeyLeft = constants.TESTRPC_PRIVATE_KEYS[accounts.indexOf(makerAddressLeft)];
+        orderFactoryLeft = new OrderFactory(privateKeyLeft, defaultOrderParamsLeft);
+        const privateKeyRight = constants.TESTRPC_PRIVATE_KEYS[accounts.indexOf(makerAddressRight)];
+        orderFactoryRight = new OrderFactory(privateKeyRight, defaultOrderParamsRight);
+    });
+
+    blockchainTests('assertValidMatch', () => {
+        it('should revert if the prices of the left order is less than the price of the right order', async () => {
+            const leftOrder = await orderFactoryLeft.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(49, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+            });
+            const rightOrder = await orderFactoryRight.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+            });
+            const orderHashHexLeft = orderHashUtils.getOrderHashHex(leftOrder);
+            const orderHashHexRight = orderHashUtils.getOrderHashHex(rightOrder);
+            const expectedError = new ExchangeRevertErrors.NegativeSpreadError(orderHashHexLeft, orderHashHexRight);
+            return expect(testExchange.assertValidMatch.callAsync(leftOrder, rightOrder)).to.revertWith(expectedError);
+        });
+
+        it('should succeed if the prices of the left and right orders are equal', async () => {
+            const leftOrder = await orderFactoryLeft.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+            });
+            const rightOrder = await orderFactoryRight.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+            });
+            return expect(testExchange.assertValidMatch.callAsync(leftOrder, rightOrder)).to.be.fulfilled('');
+        });
+
+        it('should succeed if the price of the left order is higher than the price of the right', async () => {
+            const leftOrder = await orderFactoryLeft.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+            });
+            const rightOrder = await orderFactoryRight.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+            });
+            return expect(testExchange.assertValidMatch.callAsync(leftOrder, rightOrder)).to.be.fulfilled('');
+        });
+    });
```

Comment:
How about tests for the two cases where `_assertValidMatch()` can overflow?

Reply:
I'm glad you mentioned this. I didn't write these intentionally because these cases should (and are) covered in the unit tests for `_safeMul`. This said, if it makes us feel better, I'm happy to add them back in.

Reply:
It's kind of more to protect against regressions, i.e., that SafeMath is still being used if we refactor this function.

Reply:
That's a fair point. I'll go ahead and change it. This said, I wonder if it makes sense to enforce a linter rule that normal arithmetic can't be used. I think that there is a rule in `solhint` that would enforce this. Do you think that would be sufficient to prevent regressions?

Reply:
I can dig it.